File: app/services/AuthorService.scala
package services

import scala.slick.driver.MySQLDriver.simple._
import models.Author
import models.daos.slick.DBTableDefinitions._
import models.daos.slick.DBTableDefinitions.{Author => AuthorTable}
import scala.slick.jdbc.StaticQuery
import scala.slick.jdbc.StaticQuery.interpolation
import scala.slick.jdbc.{GetResult, StaticQuery => Q}


class AuthorService(val authors:TableQuery[AuthorTable], val bookAuthors:TableQuery[BookAuthor], db: Database) {
    
	def createAuthor(authorData: Author):Int = {
	    
	    db.withSession { implicit session =>                  
	        val authorId:Int = (authors returning authors.map(_.id)) += createAuthorRowForInsert(authorData)

	        authorId
        }
	}
	
	def updateAuthor(authorId: Int, authorData: Author) = {
	    db.withSession { implicit session => 
	        authors.where(_.id === authorId).update(createAuthorRowForUpdate(authorId, authorData))
	    }
	}
	
	def searchAuthors(term:String):Seq[(Int, String, String)] = {
	    db.withSession { implicit session =>            
	        val query = for {
			        author <- authors if author.firstname.like(term + "%") || author.lastname.like(term + "%")
			    } yield (author.id, author.firstname, author.lastname)
			
			    query.run
      }
	}
	
	def getAuthor(authorId:Int):Option[AuthorRow] = {
	    db.withSession { implicit session => 
	        authors.filter{ _.id === authorId }.firstOption
	    }
	}
	
	def createAuthorRowForInsert(authorData:Author):AuthorRow = {
        createAuthorRow(0, authorData)
  	}
	
	def createAuthorRowForUpdate(authorId:Int, authorData:Author):AuthorRow = {
        createAuthorRow(authorId, authorData)
  	}
		
	def createAuthorRow(authorId:Int, authorData:Author):AuthorRow = {
        AuthorRow(
        	authorId,
        	authorData.firstname,
        	authorData.lastname
        )
	}

    def getAuthors(limit: Int = 0): List[(Int, String, String, Int)] = {
        val processedLimit:Int = getLimitValue(limit)

        val query = sql"""
            SELECT
              a.id,
              a.firstname,
              a.lastname,
              count(b.id) AS amount
            FROM
              book b LEFT JOIN book_author ba ON b.id = ba.book_id
              RIGHT JOIN author a ON a.id = ba.author_id
            WHERE
              a.firstname IS NOT NULL AND a.lastname IS NOT NULL
            GROUP BY
              a.id
            ORDER BY
              lastname ASC, firstname ASC
            LIMIT $processedLimit"""
            .as[(Int, String, String, Int)]

        db.withSession { implicit session =>
            query.list
        }
    }

    def getLimitValue(limit: Int): Int = {
        limit match {
            case x if x <= 0 => 100000
            case _ => limit
        }
    }
}

File: src/components/service/AuthorService.js


export default class AuthorService {
    constructor(db) {
        // maybe not needed?
        this.db = db;
    }

    setConnection(connection) {
        this.connection = connection;
    }

    getAllAuthors(callback) {
        this.connection.query(
            this.createGetAuthorsQuery(),
            callback
        );
    }

    getAuthor(id, callback) {
        this.connection.query(
            `SELECT
                a.id,
                a.firstname,
                a.lastname,
                CONCAT(firstname, ' ', lastname) AS name
            FROM 
                author AS a
            WHERE
                a.id = :id`,
            {id: id},
            callback
        );
    }  

    getAuthorWithBooks(id, callback) {
        this.connection.query(
            `SELECT
                a.id,
                a.firstname,
                a.lastname,
                CONCAT(firstname, ' ', lastname) AS name,
                b.id AS book_id,
                b.title,
                b.is_read
            FROM 
                book AS b
                LEFT JOIN book_author AS ba ON b.id = ba.book_id
                RIGHT JOIN author AS a ON a.id = ba.author_id
            WHERE
                a.id = :id`,
            {id: id},
            callback
        );
    }  

    createAuthor(author, callback) {
        this.connection.query(
            'INSERT INTO author (firstname, lastname) VALUES (:firstname, :lastname)',
            {
                firstname: author.firstname,
                lastname: author.lastname
            },
            callback
        );
    }

    updateAuthor(id, author, callback) {
        this.connection.query(
            'UPDATE author set firstname = :firstname, lastname = :lastname WHERE id = :id',
            {
                firstname: author.firstname,
                lastname: author.lastname,
                id: id
            },
            callback
        );
    }


    searchAuthors(term, callback) {
        let query = `SELECT
              a.id,
              a.firstname,
              a.lastname,
              CONCAT(firstname, ' ', lastname) AS name
            FROM
              author a
            WHERE
              a.firstname LIKE :term
              OR a.lastname LIKE :term 
            ORDER BY
              a.lastname ASC, a.firstname ASC`;

        this.connection.query(query, {term: term + '%'}, callback);
    }

    createGetAuthorsQuery() {
        let query = `SELECT
              a.id,
              a.firstname,
              a.lastname,
              CONCAT(firstname, ' ', lastname) AS name,
              count(b.id) AS amount
            FROM
              book b LEFT JOIN book_author ba ON b.id = ba.book_id
              RIGHT JOIN author a ON a.id = ba.author_id
            WHERE
              a.firstname IS NOT NULL AND a.lastname IS NOT NULL
            GROUP BY
              a.id
            ORDER BY
              lastname ASC, firstname ASC`;
        
        return query;
    }

    createAuthorObject(result) {
        let author = {};

        result.map(row => {
            if (null == author.id) {
                author.id = row['id'];
                author.firstname = row['firstname'];
                author.lastname = row['lastname'];
                author.name = row['name'];
            }

            if (null == author.books) {
                author.books = [];
            }

            if (null !== row['book_id']) {
                author.books.push({
                    id: row['book_id'],
                    title: row['title'],
                    is_read: row['is_read'],
                });                
            }

            author.amount = author.books.length;
        });

        return author;
    }
};
